chips/sifive/src/plic.rs

```diff
@@ -69,15 +89,48 @@ impl Plic {
         }
     }
 
+    /// Enable a list of interrupt IDs. The IDs must be in the range 1..TOTAL_INTS.
+    pub fn enable_specific_interrupts(&self, interrupts: &[u32]) {
+        for interrupt in interrupts {
+            let enable_regs = &self.registers.enable[0..TOTAL_INTS.div_ceil(32)];
+            let priority_regs = &self.registers.priority[0..TOTAL_INTS];
+            let offset = interrupt / 32;
+            let irq = interrupt % 32;
+            let old_value = enable_regs[offset as usize].get();
+            enable_regs[offset as usize].set(old_value | (1 << irq));
+
+            // Set some default priority for each interrupt. This is not really used
+            // at this point.
+            // The priority registers indexed 0 for interrupt 1, 1 for interrupt 2, etc.
+            // so we subtract 1 from the interrupt number to get the correct index.
+            priority_regs[*interrupt as usize - 1].write(priority::Priority.val(4));
+        }
+        // Accept all interrupts.
+        self.registers.threshold.write(priority::Priority.val(0));
+    }
+
+    pub fn disable_specific_interrupts(&self, interrupts: &[u32]) {
+        let enable_regs = &self.registers.enable[0..TOTAL_INTS.div_ceil(32)];
+        for interrupt in interrupts {
+            let offset = interrupt / 32;
+            let irq = interrupt % 32;
+            let old_value = enable_regs[offset as usize].get();
+            enable_regs[offset as usize].set(old_value & !(1 << irq));
+        }
+    }
+
     /// Enable all interrupts.
     pub fn enable_all(&self) {
-        for enable in self.registers.enable.iter() {
+        let enable_regs = &self.registers.enable[0..TOTAL_INTS.div_ceil(32)];
+        let priority_regs = &self.registers.priority[0..TOTAL_INTS];
```

Comment:
Why does `enable_regs` need the div, but priority doesn't? [I guess `enable` is bit-banded mapping, but priority is one-per reg?]. This is really subtle, and requires somewhat deep knowledge of risc-v to get correct.

I think if we _have_ to have something like this, I would rather see a type-wrapper or similar so that `iter()` continues to work correctly, and the nuance of `TOTAL_INTS` can be contained to one place.

Reply:
> I think if we have to have something like this, I would rather see a type-wrapper or similar so that iter() continues to work correctly, and the nuance of TOTAL_INTS can be contained to one place.

In fact after looking into it a bit, I am not sure how to go about it. Each and every time I am hitting the wall that the generic `const` parameter cannot be used in `const` expressions. Making a type wrapper around the arrays of registers with extra fields that *could* be initialized in runtime is also a no-go, since this time needs to be transparent. Could you elaborate what kind of approach you were thinking about?